File: Sources/Elastos/Frameworks/Droid/DevSamples/Gallery/src/data/AsyncImageLoader.cpp
#include "AsyncImageLoader.h"
#include <elastos/utility/logging/Logger.h>
#include <elastos/core/Math.h>
#include <unistd.h>
#include <elastos/core/StringBuilder.h>

using Elastos::Core::EIID_IRunnable;
using Elastos::Core::CThread;
using Elastos::Core::IThread;
using Elastos::Core::StringBuilder;
using Elastos::IO::CFile;
using Elastos::IO::IFile;
using Elastos::Droid::Graphics::CBitmapFactory;
using Elastos::Droid::Graphics::CBitmapFactoryOptions;
using Elastos::Droid::Graphics::IBitmapFactory;
using Elastos::Droid::Graphics::Drawable::CBitmapDrawable;
using Elastos::Droid::Graphics::Drawable::EIID_IBitmapDrawable;
using Elastos::Utility::Logging::Logger;

namespace Gallery {

const String AsyncImageLoader::TAG("AsyncImageLoader");

Int32 AsyncImageLoader::sImgWidth = 320;
Int32 AsyncImageLoader::sImgHeight = 320;
AutoPtr<ImageCacheService> AsyncImageLoader::sCacheService;
HashMap<String, AutoPtr<IBitmapDrawable> > AsyncImageLoader::sImageCache;
HashMap<String, Boolean> AsyncImageLoader::sImageSampleStatus;
HashMap<String, AutoPtr<IWeakReference> > AsyncImageLoader::sHighImageCache;
// HashMap<String, AutoPtr<IBitmapDrawable> > AsyncImageLoader::sHighImageCache;
HashMap<String, AutoPtr< ArrayOf<Int32> > > AsyncImageLoader::sHighImageWH;
HashMap<String, AutoPtr<List<AutoPtr<ImageLoaderParam> > > > AsyncImageLoader::sImageLoaderParamCache;
Mutex AsyncImageLoader::sLock;
Mutex AsyncImageLoader::sStatusLock;
Mutex AsyncImageLoader::sImageWHLock;

AsyncImageLoader::ImageLoaderThread::ImageLoaderThread(
    /* [in] */ Boolean isHigh,
    /* [in] */ const String& imageUrl)
    : mIsHigh(isHigh)
    , mImageUrl(imageUrl)
{}

CAR_INTERFACE_IMPL(AsyncImageLoader::ImageLoaderThread, IRunnable)

ECode AsyncImageLoader::ImageLoaderThread::Run()
{
    Logger::D(TAG, "ImageLoaderThread::++++ tid:%d", gettid());
    AutoPtr<IBitmapDrawable> drawable = LoadImageFromUrl(mImageUrl);
    AsyncImageLoader::DrawableLoaded(mImageUrl, mIsHigh, drawable);

    Logger::D(TAG, "ImageLoaderThread::----- drawable:%p tid:%d", drawable.Get(), gettid());
    return NOERROR;
}

AutoPtr<IBitmapDrawable> AsyncImageLoader::ImageLoaderThread::LoadImageFromUrl(
    /* [in] */ const String& path)
{
    Logger::D(TAG, "LoadImageFromUrl()----path:%s tid:%d", path.string(), gettid());
    AutoPtr<ImageCacheService> cacheService = GetCacheService();
    AutoPtr<IBitmap> bitmap;
    Int64 last = 0;
    AutoPtr<IFile> file;
    CFile::New(path, (IFile**)&file);
    file->LastModified(&last);
    Int32 hashkey = path.GetHashCode();
    Int64 len = 0;
    file->GetLength(&len);
    cacheService->GetImageData(hashkey, mIsHigh, len, last, (IBitmap**)&bitmap);
    if (bitmap != NULL) {
        {
            AutoLock lock(sStatusLock);
            sImageSampleStatus[path] = TRUE;
        }
        AutoPtr<IBitmapDrawable> bitmapDrawable;
        CBitmapDrawable::New(bitmap, (IBitmapDrawable**)&bitmapDrawable);
        Logger::D(TAG, "LoadImageFromUrl()---1---bitmapDrawable:%p", bitmapDrawable.Get());
        return bitmapDrawable;
    }

    AutoPtr<IBitmapFactoryOptions> opt;
    Int32 inSampleSize = 0;
    CBitmapFactoryOptions::New((IBitmapFactoryOptions**)&opt);
    opt->SetInJustDecodeBounds(TRUE);
    AutoPtr<IBitmapFactory> factory;
    CBitmapFactory::AcquireSingleton((IBitmapFactory**)&factory);
    factory->DecodeFile(path, opt, (IBitmap**)&bitmap);

    if (mIsHigh) {
        inSampleSize = ComputeSampleSize(opt, -1, 1280 * 672 * 4);
    }
    else {
        inSampleSize = ComputeSampleSize(opt, -1, sImgWidth * sImgHeight);
    }

    Logger::D(TAG, "LoadImageFromUrl()----inSampleSize:%d tid:%d",
        inSampleSize, gettid());
    opt->SetInSampleSize(inSampleSize);
    opt->SetInJustDecodeBounds(FALSE);

    {
        AutoLock lock(sStatusLock);
        sImageSampleStatus[path] = (inSampleSize > 1 || mIsHigh);
    }

    bitmap = NULL;
    factory->DecodeFile(path, opt, (IBitmap**)&bitmap);
    AutoPtr<IBitmapDrawable> bitmapDrawable;
    CBitmapDrawable::New(bitmap, (IBitmapDrawable**)&bitmapDrawable);
    if (inSampleSize > 1 && bitmap != NULL) {
        cacheService->PutImageData(hashkey, mIsHigh, len, last, bitmap);
    }
    Logger::D(TAG, "LoadImageFromUrl()---2---bitmapDrawable:%p tid:%d",
        bitmapDrawable.Get(), gettid());
    return bitmapDrawable;
}

Int32 AsyncImageLoader::ImageLoaderThread::ComputeSampleSize(
    /* [in] */ IBitmapFactoryOptions* options,
    /* [in] */ Int32 minSideLength,
    /* [in] */ Int32 maxNumOfPixels)
{
    Int32 initialSize = ComputeInitialSampleSize(options, minSideLength, maxNumOfPixels);
    Int32 roundedSize = 0;

    if (initialSize <= 8 ) {
        roundedSize = 1;
        while (roundedSize < initialSize) {
            roundedSize <<= 1;
        }
    }
    else {
        roundedSize = (initialSize + 7) / 8 * 8;
    }

    return roundedSize;
}

Int32 AsyncImageLoader::ImageLoaderThread::ComputeInitialSampleSize(
    /* [in] */ IBitmapFactoryOptions* options,
    /* [in] */ Int32 minSideLength,
    /* [in] */ Int32 maxNumOfPixels)
{
    Int32 w = 0;
    Int32 h = 0;
    options->GetOutWidth(&w);
    options->GetOutHeight(&h);
    Logger::D(TAG, "ComputeInitialSampleSize----w:%d,h:%d, tid:%d", w, h, gettid());

    if (mIsHigh) {
        AutoLock lock(sImageWHLock);
        AutoPtr< ArrayOf<Int32> > origionWH = ArrayOf<Int32>::Alloc(2);
        (*origionWH)[0] = w;
        (*origionWH)[1] = h;
        sHighImageWH[mImageUrl] = origionWH;

        Logger::D(TAG, "ComputeInitialSampleSize----size:%d", sHighImageWH.GetSize());
    }

    Int32 lowerBound = (maxNumOfPixels == -1/*UNCONSTRAINED*/) ? 1 :
        (Int32) Elastos::Core::Math::Ceil(Elastos::Core::Math::Sqrt((Double)w * h / maxNumOfPixels));
    Int32 upperBound = (minSideLength == -1/*UNCONSTRAINED*/) ? 128 :
        (Int32) Elastos::Core::Math::Min(Elastos::Core::Math::Floor((Double)w / minSideLength),
        Elastos::Core::Math::Floor((Double)(h / minSideLength)));

    if (upperBound < lowerBound) {
        // return the larger one when there is no overlapping zone.
        return lowerBound;
    }

    if ((maxNumOfPixels == -1/*UNCONSTRAINED*/) && (minSideLength == -1/*UNCONSTRAINED*/)) {
        return 1;
    }
    else if (minSideLength == -1/*UNCONSTRAINED*/) {
        return lowerBound;
    }
    else {
        return upperBound;
    }
}

AutoPtr<ImageCacheService> AsyncImageLoader::GetCacheService()
{
    if (sCacheService == NULL) {
        sCacheService = new ImageCacheService(sImgWidth, sImgHeight);
    }
    return sCacheService;
}

String AsyncImageLoader::MakeImageParamKey(
    /* [in] */ const String& imageUrl,
    /* [in] */ Boolean isHigh)
{
    StringBuilder sb(imageUrl);
    sb +="isHigh=";
    sb += isHigh;
    return sb.ToString();
}

AutoPtr<IBitmapDrawable> AsyncImageLoader::LoadDrawable(
    /* [in] */ const String& imageUrl,
    /* [in] */ Boolean isHigh,
    /* [in] */ IImageView* imageView,
    /* [in] */ ILoadImageCallback* imageCallback)
{
    Logger::D(TAG, "===================================================");
    Logger::D(TAG, "LoadDrawable()--imageUrl:%s,isHigh:%d, tid:%d", imageUrl.string(), isHigh, gettid());

    AutoLock lock(sLock);

    if (isHigh) {
        HashMap<String, AutoPtr<IWeakReference> >::Iterator it = sHighImageCache.Find(imageUrl);
        // HashMap<String, AutoPtr<IBitmapDrawable> >::Iterator it = sHighImageCache.Find(imageUrl);
        if (it != sHighImageCache.End()) {
            AutoPtr<IBitmapDrawable> drawable;
            AutoPtr<IWeakReference> drawableRef = it->mSecond;
            drawableRef->Resolve(EIID_IBitmapDrawable, (IInterface**)&drawable);
            // AutoPtr<IBitmapDrawable> drawable = it->mSecond;
            if (drawable != NULL) {
                return drawable;
            }
        }
    }
    else {
        HashMap<String, AutoPtr<IBitmapDrawable> >::Iterator it = sImageCache.Find(imageUrl);
        if (it != sImageCache.End()) {
            AutoPtr<IBitmapDrawable> drawable = it->mSecond;
            if (drawable != NULL) {
                return drawable;
            }
        }
    }

    {
        String key = MakeImageParamKey(imageUrl, isHigh);
        HashMap<String, AutoPtr<List<AutoPtr<ImageLoaderParam> > > >::Iterator it = sImageLoaderParamCache.Find(key);
        if (it != sImageLoaderParamCache.End()) {
            Logger::D(TAG, "===========It's already on loading process...");
            Boolean needAdd = TRUE;
            AutoPtr<List<AutoPtr<ImageLoaderParam> > > list = it->mSecond;
            List<AutoPtr<ImageLoaderParam> >::Iterator pit = list->Begin();
            for (; pit != list->End(); ++pit) {
                ImageLoaderParam* param = *pit;
                if (param->mImageView.Get() == imageView) {
                    Logger::D(TAG, "===========It's loading image for same imageView...");
                    needAdd = FALSE;
                    break;
                }
            }

            if (needAdd) {
                Logger::D(TAG, "===========It's loading image for new imageView...");
                AutoPtr<ImageLoaderParam> param = new ImageLoaderParam();
                param->mImageView = imageView;
                param->mCallback = imageCallback;
                list->PushBack(param);
            }
        }
        else {
            Logger::D(TAG, "===========It's a new loading process..., tid: %d", gettid());

            AutoPtr< List<AutoPtr<ImageLoaderParam> >  > list = new List<AutoPtr<ImageLoaderParam> >();
            AutoPtr<ImageLoaderParam> param = new ImageLoaderParam();
            param->mImageView = imageView;
            param->mCallback = imageCallback;
            list->PushBack(param);
            sImageLoaderParamCache[key] = list;

            AutoPtr<IThread> loadThread;
            AutoPtr<ImageLoaderThread> loadRunnable = new ImageLoaderThread(isHigh, imageUrl);
            CThread::New(IRunnable::Probe(loadRunnable), (IThread**)&loadThread);
            assert(loadThread != NULL);
            loadThread->Start();
        }
    }

    return NULL;
}

void AsyncImageLoader::DrawableLoaded(
    /* [in] */ const String& imageUrl,
    /* [in] */ Boolean isHigh,
    /* [in] */ IBitmapDrawable* d)
{
    AutoLock lock(sLock);

    AutoPtr<IBitmapDrawable> drawable = d;
    if (isHigh) {
        AutoPtr<IWeakReference> drawableRef;
        IWeakReferenceSource::Probe(drawable)->GetWeakReference((IWeakReference**)&drawableRef);
        sHighImageCache[imageUrl] = drawableRef;
        // sHighImageCache[imageUrl] = drawable;
        Logger::D(TAG, "ImageLoaderThread::----sHighImageCache---size:%d tid:%d",
            sHighImageCache.GetSize(), gettid());
    }
    else {
        sImageCache[imageUrl] = drawable;
        Logger::D(TAG, "ImageLoaderThread::----sImageCache---size:%d tid:%d",
            sImageCache.GetSize(), gettid());
    }

    {
        String key = MakeImageParamKey(imageUrl, isHigh);
        HashMap<String, AutoPtr<List<AutoPtr<ImageLoaderParam> > > >::Iterator it = sImageLoaderParamCache.Find(key);
        if (it != sImageLoaderParamCache.End()) {
            AutoPtr<List<AutoPtr<ImageLoaderParam> > > list = it->mSecond;
            List<AutoPtr<ImageLoaderParam> >::Iterator pit = list->Begin();
            for (; pit != list->End(); ++pit) {
                AutoPtr<ImageLoaderParam> param = *pit;
                param->mCallback->ImageLoaded(drawable, param->mImageView);
            }

            list->Clear();
            sImageLoaderParamCache.Erase(it);
        }
    }
}

Boolean AsyncImageLoader::NeedLoadHighDrawable(
    /* [in] */ const String& imageUrl)
{
    AutoLock lock(sStatusLock);
    HashMap<String, Boolean>::Iterator it = sImageSampleStatus.Find(imageUrl);
    if (it != sImageSampleStatus.End()) {
        return it->mSecond;
    }

    return TRUE;
}

Boolean AsyncImageLoader::GetOrigionWidthAndHeight(
    /* [in] */ const String& imageUrl,
    /* [out] */ Int32* w,
    /* [out] */ Int32* h)
{
    // VALIDATE_NOT_NULL(w);
    *w = 0;
    // VALIDATE_NOT_NULL(h);
    *h = 0;

    AutoLock lock(sImageWHLock);
    HashMap<String, AutoPtr< ArrayOf<Int32> > >::Iterator it = sHighImageWH.Find(imageUrl);
    if (it != sHighImageWH.End()) {
        AutoPtr< ArrayOf<Int32> > tmpArray = it->mSecond;
        *w = (*tmpArray)[0];
        *h = (*tmpArray)[1];
        return TRUE;
    }
    AutoPtr<IBitmapFactoryOptions> opt;
    CBitmapFactoryOptions::New((IBitmapFactoryOptions**)&opt);
    opt->SetInJustDecodeBounds(TRUE);
    AutoPtr<IBitmapFactory> factory;
    CBitmapFactory::AcquireSingleton((IBitmapFactory**)&factory);
    AutoPtr<IBitmap> bitmap;
    if (FAILED(factory->DecodeFile(imageUrl, opt, (IBitmap**)&bitmap)))
        return FALSE;
    opt->GetOutWidth(w);
    opt->GetOutHeight(h);
    return TRUE;
}

void AsyncImageLoader::ClearAll()
{
    Logger::D(TAG, "ClearAll---");
    sImageCache.Clear();
    sImageSampleStatus.Clear();
    sHighImageCache.Clear();
    sHighImageWH.Clear();
}

void AsyncImageLoader::UpdateImageCache(
    /* [in] */ ArrayOf<String>* albumArray)
{
    Logger::D(TAG, "UpdateImageCache()---");
    AutoPtr<ImageCacheService> cacheService = GetCacheService();
    AutoPtr< ArrayOf<String> > album = albumArray;
    cacheService->UpdateCache();
    String albumPath;
    String key;
    Int32 len = album->GetLength();
    Boolean isMatch = FALSE;
    HashMap<String, AutoPtr<IBitmapDrawable> >::Iterator it = sImageCache.Begin();
    for (; it != sImageCache.End();) {
        key = it->mFirst;
        isMatch = FALSE;
        for (Int32 i = 0; i != len; i++) {
            if (key.Equals((*album)[i])) {
                isMatch = TRUE;
                break;
            }
        }
        if (isMatch) {
            ++it;
        }
        else {
            sImageCache.Erase(it++);
        }
    }
}

} // namespace Gallery

File: Sources/Elastos/Frameworks/Droid/DevSamples/jpk/Common/js/Gallery/data/AsyncImageLoader.js
module.exports = function(aoElastos){

var JS_MULTI_READY = false;

//common definition
var CString = aoElastos.Core.CString;
var Droid_New = aoElastos.Droid.New;
var Core_New = aoElastos.Core.New;

function extend (a,b) {
    for (var p in b) a[p] = b[p];
}

var sFileName = '/data/temp/node/JSGallery/data/ImageCacheService.js';
var ImageCacheService = require(sFileName)(aoElastos);

//main class
function AsyncImageLoader () {};

//private static property:
var TAG = "AsyncImageLoader";
//var sImgWidth = 320;
//var sImgHeight = 320;
var sImgWidth = 952;
var sImgHeight = 1648;

var sCacheService=null;

var sImageCache = {};
var sImageSampleStatus = {};
var sHighImageCache = {};
var sHighImageWH = {};
var sImageLoaderParamCache = {};

var sLock = null;
var sStatusLock = null;
var sImageWHLock = null;

//class
function ILoadImageCallback () {
    this.ImageLoaded = function (aoImageDrawable, aoImageView) {
        ///* [in] */ IDrawable* imageDrawable,
        ///* [in] */ IImageView* imageView) = 0;
    };
}

//class
function ImageLoaderParam (aoImageView, aoCallback) {
    this.mImageView = aoImageView;  //IImageView
    this.mCallback = aoCallback;    //ILoadImageCallback
}

//private internal class
var ImageLoaderThread = (function(){
    function _ImageLoaderThread (abIsHigh, asImageUrl) {
        this.mIsHigh = abIsHigh;
        this.mImageUrl = asImageUrl;
    }

    //private method
    function LoadImageFromUrl (path) {
        var cacheService = GetCacheService();
        var last = 0;
        var file = Core_New("Elastos.IO.CFile", path);
        //var last = file.LastModified();
        var last = file.GetLastModified();

        //var hashkey = path.GetHashCode();
        var hashkey = file.GetHashCode();
        if (hashkey < 0) hashkey = 0 - hashkey;
        var len = file.GetLength();
        var bitmap = cacheService.GetImageData(hashkey, this.mIsHigh, len, last);
        if (bitmap) {
            //Mutex::Autolock lock(sStatusLock);
            sImageSampleStatus[path] = true;
            var bitmapDrawable = Droid_New("Elastos.Droid.Graphics.Drawable.CBitmapDrawable",bitmap);

            return bitmapDrawable;
        }

        var opt;
        var inSampleSize = 0;

        var opt = Droid_New("Elastos.Droid.Graphics.CBitmapFactoryOptions");
        opt.SetInJustDecodeBounds(true);

        var factory = Droid_New("Elastos.Droid.Graphics.CBitmapFactory");

        var bitmap = factory.DecodeFile(path, opt);

        if (this.mIsHigh) {
            //inSampleSize = ComputeSampleSize(opt, -1, 1280 * 672 * 4);
            inSampleSize = ComputeSampleSize(opt, -1, 1648 * 952 * 4);
        }
        else {
            inSampleSize = ComputeSampleSize(opt, -1, sImgWidth * sImgHeight);
        }

        opt.SetInSampleSize(inSampleSize);
        opt.SetInJustDecodeBounds(false);

        //Mutex::Autolock lock(sStatusLock);
        sImageSampleStatus[path] = (inSampleSize > 1 || this.mIsHigh);

        //bitmap = null;
        bitmap = factory.DecodeFile(path, opt);

        var bitmapDrawable = Droid_New("Elastos.Droid.Graphics.Drawable.CBitmapDrawable",bitmap);
        if (inSampleSize > 1 && bitmap != null) {
            cacheService.PutImageData(hashkey, this.mIsHigh, len, last, bitmap);
        }

        return bitmapDrawable;
    }

    function ComputeSampleSize (options, minSideLength, maxNumOfPixels) {
        var initialSize = ComputeInitialSampleSize(options, minSideLength, maxNumOfPixels);
        var roundedSize = 0;

        if (initialSize <= 8 ) {
            roundedSize = 1;
            while (roundedSize < initialSize) {
                roundedSize <<= 1;
            }
        }
        else {
            roundedSize = (initialSize + 7) / 8 * 8;
        }

        return roundedSize;
    }

    function ComputeInitialSampleSize (options, minSideLength, maxNumOfPixels) {
        var w = options.GetOutWidth();
        var h = options.GetOutHeight();

        if (this.mIsHigh) {
            //Mutex::Autolock lock(sImageWHLock);
            sHighImageWH[mImageUrl] = [w, h];
        }

        var lowerBound = (maxNumOfPixels == -1/*UNCONSTRAINED*/) ? 1 :
            Math.ceil(Math.sqrt(w * h / maxNumOfPixels));
        var upperBound = (minSideLength == -1/*UNCONSTRAINED*/) ? 128 :
            Math.min(Math.Floor(w / minSideLength),
                Math.floor(h / minSideLength));

        if (upperBound < lowerBound) {
            // return the larger one when there is no overlapping zone.
            return lowerBound;
        }

        if ((maxNumOfPixels == -1/*UNCONSTRAINED*/) && (minSideLength == -1/*UNCONSTRAINED*/)) {
            return 1;
        }
        else if (minSideLength == -1/*UNCONSTRAINED*/) {
            return lowerBound;
        }
        else {
            return upperBound;
        }

    }

    //public method:
    extend(_ImageLoaderThread.prototype, {
        _LoadImageFromUrl : LoadImageFromUrl,

        start : function () {
            if(!JS_MULTI_READY) {
                return this.Run();
            }
            else {
                this.Run();
            }
        },

        Run : function () {
            var drawable = this._LoadImageFromUrl(this.mImageUrl);

            AsyncImageLoader.DrawableLoaded(this.mImageUrl, this.mIsHigh, drawable);

            if(!JS_MULTI_READY) {
                return drawable;
            }
        },
    }); //extend

    return _ImageLoaderThread;
})();   //ImageLoaderThread

//public static
function LoadDrawable (imageUrl, isHigh, imageView, imageCallback) {
    var drawable;

    //Mutex::Autolock lock(sLock);

    if (isHigh) {
        drawable = sHighImageCache[imageUrl];
    }
    else {
        drawable = sImageCache[imageUrl];
    }
    if (drawable) return drawable;

    var key = this.MakeImageParamKey(imageUrl, isHigh);
    var list = sImageLoaderParamCache[key];
    if (list) {
        var needAdd = true;
        for (var i=0, im=list.length; i<im; i++) {
            if (list[i] == imageView) {
                needAdd = false;
                break;
            }
        }
        if (needAdd) {
            list.push({
                mImageView : imageView,
                mCallback : imageCallback,
            });
        }
    }
    else {
        sImageLoaderParamCache[key] = [
            {
                mImageView : imageView,
                mCallback : imageCallback,
            }
        ];
    }

    var loadRunnable = new ImageLoaderThread(isHigh, imageUrl);

    if(JS_MULTI_READY) {
        loadRunnable.start();
    }
    else {
        drawable = loadRunnable.start();
    }

    return drawable;
}

function DrawableLoaded (imageUrl, isHigh, drawable){
    //Mutex::Autolock lock(sLock);

    if (isHigh) {
        sHighImageCache[imageUrl] = drawable;
    }
    else {
        sImageCache[imageUrl] = drawable;
    }

    var key = this.MakeImageParamKey(imageUrl, isHigh);
    var list = sImageLoaderParamCache[key];
    if (list) {
        for (var i=0, im=list.length; i<im; i++) {
            var param = list[i];    //ImageLoaderParam
            param.mCallback.ImageLoaded(drawable, param.mImageView);
        }
    }
}

function NeedLoadHighDrawable (imageUrl) {
    //Mutex::Autolock lock(sStatusLock);
    //return sImageSampleStatus[imageUrl];
    return true;
}

// Boolean AsyncImageLoader::GetOrigionWidthAndHeight(
function GetOrigionWidthAndHeight(
//     /* [in] */ const String& imageUrl,
        imageUrl,
//     /* [out] */ Int32* w,
//     /* [out] */ Int32* h)
        r)  //r:{w:w,h:h}   more than one out para, must transfer throught object reference
{

//     // VALIDATE_NOT_NULL(w);
//     *w = 0;
        r.w = 0;
//     // VALIDATE_NOT_NULL(h);
//     *h = 0;
        r.h = 0;

//     Mutex::Autolock lock(sImageWHLock);
        //TODO:multi thread opt
//     HashMap<String, AutoPtr< ArrayOf<Int32> > >::Iterator it = sHighImageWH.Find(imageUrl);
        var it = sHighImageWH[imageUrl];
//     if (it != sHighImageWH.End()) {
        if (it) {
//         AutoPtr< ArrayOf<Int32> > tmpArray = it->mSecond;
            var tmpArray = it;
//         *w = (*tmpArray)[0];
            r.w = tmpArray[0];
//         *h = (*tmpArray)[1];
            r.h = tmpArray[1];
//         return TRUE;
            return true;
//     }
        }
//     AutoPtr<IBitmapFactoryOptions> opt;
        var opt;

//     CBitmapFactoryOptions::New((IBitmapFactoryOptions**)&opt);
        opt = Droid_New("Elastos.Droid.Graphics.CBitmapFactoryOptions");
//     opt->SetInJustDecodeBounds(TRUE);
        opt.SetInJustDecodeBounds(true);
//     AutoPtr<IBitmapFactory> factory;
        var factory;
//     CBitmapFactory::AcquireSingleton((IBitmapFactory**)&factory);
        factory = Droid_New("Elastos.Droid.Graphics.CBitmapFactory");
//     AutoPtr<IBitmap> bitmap;
        var bitmap;
//     if (FAILED(factory->DecodeFile(imageUrl, opt, (IBitmap**)&bitmap)))
//         return FALSE;
        try {
            bitmap = factory.DecodeFile(imageUrl, opt);
        }
        catch(e) {
            return false;
        }
//     opt->GetOutWidth(w);
        r.w = opt.GetOutWidth();
//     opt->GetOutHeight(h);
        r.h = opt.GetOutHeight();
//     return TRUE;
        return true;
}

function ClearAll () {
    sImageCache = {};
    sImageSampleStatus = {};
    sHighImageCache = {};
    sHighImageWH = {};
}

function MakeImageParamKey (imageUrl, isHigh) {
    return imageUrl + isHigh;
}

function UpdateImageCache (albumArray) {
    var cacheService = GetCacheService();
    var album = albumArray;
    cacheService.UpdateCache();
    var len = album.length;
    var isMatch = false;
    for (var i=0,im=album.length;i<im;i++) {
        var key = album[i];
        if(sImageCache[key]){
            sImageCache[key]=null;
            delete sImageCache[key];
        }
    }
}

function GetCacheService () {
    if (sCacheService===null) {
        sCacheService = new ImageCacheService(sImgWidth, sImgHeight);
    }
    return sCacheService;
}

extend(AsyncImageLoader, {
    //public static:
    LoadDrawable : LoadDrawable,
    DrawableLoaded : DrawableLoaded,
    NeedLoadHighDrawable : NeedLoadHighDrawable,
    GetOrigionWidthAndHeight : GetOrigionWidthAndHeight,
    ClearAll : ClearAll,
    MakeImageParamKey : MakeImageParamKey,
    UpdateImageCache : UpdateImageCache,
    GetCacheService : GetCacheService,
});

return AsyncImageLoader;

};   //module.exports
